Modules/_io/fileio.c

```diff
@@ -447,6 +448,23 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
         }
     }
     else {
+#if defined(S_ISREG) && defined(ENOTDIR)
+        /* On AIX and Windows, open may succeed for files with a trailing slash.
+           The Open Group specifies filenames ending with a trailing slash should
+           be an error - ENOTDIR */
+        if (S_ISREG(fdfstat.st_mode)) {
+#ifdef MS_WINDOWS
+                rcpt = strrchr(widename, '\');
```

Comment:
For Windows use `wcslen(widename)` and `wcsrchr(widename, L'\\')`. The latter returns a `wchar_t *`. The Windows code also needs to look for forward slash.

You could separate the Windows and POSIX checks using a boolean. For example:

```C
if (S_ISREG(fdfstat.st_mode)) {
    int trailing_slash = 0;
#ifdef MS_WINDOWS
    size_t len = wcslen(widename);
    if (len && (widename[len - 1] == L'\\' || widename[len - 1] == L'/')) {
        trailing_slash = 1;
    }
#else
    size_t len = strlen(name);
    if (len && name[len - 1] == '/') {
        trailing_slash = 1;
    }
#endif
    if (trailing_slash) {
        errno = ENOTDIR;
        PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, nameobj);
        goto error;
    }
}
```